### PR TITLE
llvm 3.7.1

### PR DESCRIPTION
**Pull request by quantumsteve**
Release 3.7.1 fixes the ABI compatibility issues raised in #43499.

### CODE REVIEW

**Comment by DomT4** [comment]
`creduce` will need a `revision` bump here. Other `brew uses llvm` stuff that isn't `:optional` could do with checking locally as well, just in case.

**Comment by quantumsteve** [comment]
output from `brew uses llvm`
- [ ] cctools
- [ ] creduce
- [ ] crystal-lang
- [ ] emacs-clang-complete-async
- [ ] ldc 
- [ ] ponyc
- [ ] rtags
- [ ] rust
- [ ] homebrew/science/pocl

**Comment by DomT4** [comment]
Not all of those will be mandatory. Anything that isn't mandatory or `:recommended` doesn't need checking, and you can't do anything about the tap ones till after this is merged.

**Comment by UniqMartin** [comment]
> ponyc

I know from personal experience that this will fail. There's an [`llvm37` branch](https://github.com/CausalityLtd/ponyc/tree/llvm37) that might help, but I think the currently stable release only supports LLVM 3.6.

**Comment by John-Colvin** [comment]
ldc should be compatible with llvm 3.5 - 3.8

File: Library/Formula/llvm.rb [hunk 1]
```diff
@@ -257,7 +258,7 @@ def caveats
           return 0;
         }
       EOS
-      system "#{bin}/clang++", "test.cpp", "-o", "test"
+      system "#{bin}/clang++", "-stdlib=libstdc++", "test.cpp", "-o", "test"
```
Comment (ilovezfs) [inline Library/Formula/llvm.rb]: Should libstdc++ be used unconditionally or only when Mavericks and above?

Comment (mikemcquaid) [inline Library/Formula/llvm.rb]: I'm not sure I understand why it's being used here at all?

Comment (quantumsteve) [inline Library/Formula/llvm.rb]: I was trying to work around the following error:
```
$ brew test llvm
Testing llvm
==> Using the sandbox
==> /usr/local/Cellar/llvm/3.7.1/bin/llvm-config --prefix
==> /usr/local/Cellar/llvm/3.7.1/bin/clang++ test.cpp -o test
Last 15 lines from /Users/svh/Library/Logs/Homebrew/llvm/01.clang++:
2016-03-16 12:11:45 -0400

/usr/local/Cellar/llvm/3.7.1/bin/clang++
test.cpp
-o
test

test.cpp:1:10: fatal error: 'iostream' file not found
#include <iostream>
         ^
1 error generated.
Error: llvm: failed
Failed executing: /usr/local/Cellar/llvm/3.7.1/bin/clang++ test.cpp -o test
/usr/local/Library/Homebrew/formula.rb:1481:in `block in system'
/usr/local/Library/Homebrew/formula.rb:1418:in `open'
/usr/local/Library/Homebrew/formula.rb:1418:in `system'
/usr/local/Library/Formula/llvm.rb:267:in `block in <class:Llvm>'
/usr/local/Library/Homebrew/formula.rb:1327:in `block in run_test'
/usr/local/Library/Homebrew/extend/fileutils.rb:37:in `mktemp'
/usr/local/Library/Homebrew/formula.rb:1323:in `run_test'
/usr/local/Library/Homebrew/test.rb:28:in `block in <main>'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/timeout.rb:66:in `timeout'
/usr/local/Library/Homebrew/test.rb:27:in `<main>'
```

**Comment by quantumsteve** [comment]
To fix the test, I added the libc++ headers to the cmake variable `C_INCLUDE_DIRS`. On one computer running el capitan, I had to give the path relative to `#{MacOS.sdk_path}`. On another computer running  yosemite I had to give the absolute path. Does anyone know what might cause this?

**Comment by DomT4** [comment]
The repeated failure here is a time out. No need to keep pushing; either I or another maintainer will bump the timeout later and re-queue this manually.

**Comment by DomT4** [comment]
@BrewTestBot test this please

**Comment by dunn** [comment]
Looks like we just need a `revision` for rtags:
```
==> /usr/local/Cellar/rtags/2.1/bin/rc -c
dyld: Library not loaded: /usr/local/opt/llvm/lib/libclang.3.6.dylib
  Referenced from: /usr/local/Cellar/rtags/2.1/bin/rc
  Reason: image not found

**Comment by apjanke** [comment]
Doesn't seem like upstream `crystal-lang` is going to be updated for LLVM 3.7 compatibility soon. How about we switch it to `depends on 'homebrew/versions/llvm36'` to decouple it from the main LLVM formula? I'd guess `crystal-lang` has a lot fewer users, and it would be a shame to stall LLVM upgrades for the rest of the community on it. I know we normally prohibit mandatory dependencies on non-core formulae, but maybe worth making an exception here.

**Comment by UniqMartin** [comment]
> […], and it would be a shame to stall LLVM upgrades for the rest of the community on it. I know we normally prohibit mandatory dependencies on non-core formulae, but maybe worth making an exception here.

I think in the case of LLVM this is the more sensible thing to do. The alternative would be to drop a bunch of formulae with every LLVM release or maybe relocate them somewhere else (out of core). Few projects manage to keep track with upstream LLVM and update their code in a timely manner, when a new LLVM release arrives. Most of them seem to focus on moving their own code base forward instead of spending too many resources on staying current with LLVM. Many seem to prefer to skip a few releases (e.g. the Julia language still uses 3.3 in their stable series, but is moving to 3.7).

LLVM 3.8.0 has been released a few days ago and it would be a pity to remain stuck with 3.6.2 …

**Comment by DomT4** [comment]
Until we clean Versions up and it stops being such an insecure blob of nightmare-inducing mess I'm very :-1: on any situation where people are involuntarily suddenly dragged across to Versions without it being a deliberate choice.

If we have to go down that route I'd rather we write a custom, hacky requirement that checks LLVM version and nudges people to install Versions themselves.

**Comment by UniqMartin** [comment]
Another possibility would be to vendor a specific LLVM version into every formula that has an LLVM dependency but isn't compatible with the latest LLVM release. That would make some formulae pretty heavyweight, but decouple things a bit without dragging users into the abyss `homebrew/versions` seems to be. For the average user, the burden will be limited thanks to bottles. Just a thought, not sure if I really like this idea … but I similarly don't like us being unable to version bump the core `llvm` formula.

**Comment by ilovezfs** [comment]
Is there a reason not to have multiple llvm versions in core, with bottles available for each? It seems like the notion that there should be one and only one "llvm" in core is what's causing a problem, and that any given formula in core that uses "llvm" should be pinned to a specific version of llvm, which should also be available in core.

**Comment by DomT4** [comment]
The core has a firm policy on only shipping one version of each formula. There are exceptions, and those exceptions are wildly loathed by some maintainers and may well start to go away permanently once Versions is sorted.

There's an assumed promise we'll keep stuff in the core functional and bug-fixed, and that's a lot harder to do for multiple LLVM versions than it is 1. The support burden here is higher, and something we already struggle to keep on top of with the ~10 maintainers we have floating around.

Versions is the answer here, I'm not saying keep away from it forever, just not to lump people into a tap that's still a mess and is taking a while to comb through on moving forwards.

**Comment by ilovezfs** [comment]
@DomT4 OK, then how about an llvm tap?

**Comment by DomT4** [comment]
Still gonna have the same problem, just on a much smaller scale. The taps are a disproportionate burden on maintainers due to lower community engagement there and not having maintainers who _only_ maintain those taps. The problem is broader than LLVM, and I appreciate that's unhelpful for this issue, but moving x to y doesn't change the underlying problem.

The only way an LLVM tap is better is if there's someone in the community who is really, _really_ willing to take on looking after the LLVM formulae permanently and promising to support them on a more full-time basis than we can.

**Comment by ilovezfs** [comment]
@DomT4 I think the position you're taking doesn't make much sense. Certain formulae clearly need to build against certain versions of llvm whether we like it or not. Cross-tap dependencies are apparently contrary to policy. So unless you want to yank any formula that doesn't build against the latest version of llvm, what exactly is your plan? If it doesn't involve delivering the needed versions of llvm via binary bottle in some form, then I'd recommend just removing the formula.

**Comment by DomT4** [comment]
I'm not sure where you quite got your problem from there. I didn't dispute the idea that upstreams are always going to lag behind LLVM major releases on compatibility anywhere. I didn't dispute the idea that _eventually_ cross-tap dependencies for Versions will need to be a thing consequently. All I said is that mandatory cross-tap dependencies on Versions _right now_ is crappy crappy UX and I'd really rather avoid it.

I also pointed out moving LLVM formulae from one tap to another doesn't do anything beside hack around the underlying problem, which we'll then undo relatively shortly because nobody has volunteered to look after the separate LLVM Versions tap and adding yet another tap adds to our workload. I appreciate as you aren't a maintainer yet perhaps you've a slight blindspot on the maintenance burden but I did _try_ to elaborate on that in the most recent comment prior to this.

I'm not sure how you read any of my comment and decided I was decisively saying every formula must work with every LLVM release.

My current stance here is either:
* Vendor `llvm36` into the ~2 formulae that need it temporarily. This will be for, I hope at most, two months whilst I finish wrapping up the Versions changes.
* Use a custom `Requirement` to nudge people to install `llvm36` from Versions without doing it for them automagically when they go and `upgrade`. Given there are various problems with dependencies on cross-tap formulae in taps that aren't yet tapped I consider this the least likely to generate problems. This however may prevent bottling those ~2 formula temporarily.

In future:
* `llvm` will track whatever upstream stable is.
* Everything else lives in Versions.
* Versions becomes a default tap.
* Formulae that aren't keeping up with LLVM stable get shunted onto the last LLVM Versions formulae they support.

**Comment by ilovezfs** [comment]
@DomT4 The problem is the lack of intermediate bottles for vendored resources. I agree that creating a separate tap doesn't do anything but rearrange furniture, but neither does vendoring. I was responding to
>until we clean Versions up and it stops being such an insecure blob of nightmare-inducing mess

thinking that perhaps a separate tap would be less obnoxious to you than dragging people into Versions. Apparently, not.

I still don't think it makes sense to be vendoring compilers unless as vendored resources they're available as binary bottles, but since it's only ~2 formulae, this all sounds like much ado about nothing.

**Comment by mikemcquaid** [comment]
> I still don't think it makes sense to be vendoring compilers unless as vendored resources they're available as binary bottles, but since it's only ~2 formulae, this all sounds like much ado about nothing.

In this case it's more vendoring the library than the compiler though, right? I have no problem (or at least far less problem) with cross-tap build-time dependencies but it starts to become painful when you're managing runtime dependencies.

Longer-term I think a solution could be we just tap e.g. homebrew-versions and some other taps by default.

**Comment by ilovezfs** [comment]
>In this case it's more vendoring the library than the compiler though, right?

s/compilers/things that take a long time to build/

>Given there are various problems with dependencies on cross-tap formulae in taps that aren't yet tapped

@DomT4 @mikemcquaid I assume this is a reference to https://github.com/Homebrew/homebrew/issues/50066 (and additional issues?). As I noted in that issue, it was a regression caused by a single commit.

**Comment by mikemcquaid** [comment]
> s/compilers/things that take a long time to build/

Fair enough. Again, handling cross-tap dependencies like this is painful in terms of e.g. version bumps and maintenance. If we had an approach that used a cross-tap formula as a `resource` and then manually did the installation (so effectively the formula is always just a build-time dependency) I'd be OK with that.

**Comment by mikemcquaid** [comment]
I also just generally think software that doesn't work with the newest versions of libraries within a reasonable amount of time can be considered unmaintained and thus we don't want to keep it in Homebrew.

**Comment by ilovezfs** [comment]
>I also just generally think software that doesn't work with the newest versions of libraries within a reasonable amount of time can be considered unmaintained and thus we don't want to keep it in Homebrew.

Yep, which was the underlying reason why I'd said I think it would be acceptable to remove the formulae altogether.

>Again, handling cross-tap dependencies like this is painful in terms of e.g. version bumps and maintenance. If we had an approach that used a cross-tap formula as a resource and then manually did the installation (so effectively the formula is always just a build-time dependency) I'd be OK with that.

Yeah, non-core taps generally seem not to engender sunshine and smiles around here. Would it be feasible to enable bottling in the versions tap?

**Comment by DomT4** [comment]
> Would it be feasible to enable bottling in the versions tap?

Versions has had bottle support for absolutely yonks :smile:.

**Comment by ilovezfs** [comment]
@DomT4 Thank you. I think I had temporary blindness just now or something.

**Comment by ilovezfs** [comment]
Ah no, I was actually just looking at https://github.com/Homebrew/homebrew-versions/blob/master/allegro4.rb which is bottleless for some reason.

**Comment by DomT4** [comment]
Not everything is bottled. The formula would have had to have gone through PR at some point since bottles were made a thing, and well, a lot of Versions is dead weight untouched by contributors in a long time.

For that formula though the answer is more simply that it doesn't build on >10.8 and our CI is 10.9 and up.

**Comment by ilovezfs** [comment]
>If we had an approach that used a cross-tap formula as a resource and then manually did the installation (so effectively the formula is always just a build-time dependency) I'd be OK with that.

I wonder if having brew install target the regular Cellar but having the brew-link step link within subdirectories of the libexec directory of whichever formula for which it's a resource would do the trick.

**Comment by mikemcquaid** [comment]
@ilovezfs I think for cases like this it'd be good to have it physically install the files and duplicate them rather than symlinking.

**Comment by ilovezfs** [comment]
@mikemcquaid My concern was that the bottle paths (as dylib library paths) for non-relocatable bottles would be wrong.

**Comment by apjanke** [comment]
The legacy-homebrew repo is deprecated, and LLVM 3.8.0 is out now. Closing this as moot.

If you have time and would like to get a newer LLVM in to Homebrew, please submit a PR for `llvm` 3.8 on the new [`homebrew/core` repo](https://github.com/homebrew/homebrew-core).

**Pull request closed without merging**